### PR TITLE
LPD-37220 , LPD-35736, LPD-36796, LPD-36799, LPD-37397, LPD-37459, LPD-34556

### PR DESCRIPTION
**Pull request by NemethNorbert**
Hey Team,

[LPD-35736](https://liferay.atlassian.net/browse/LPD-35736) - Modify DatePicker to update from state change of data-engine
[LPD-36796](https://liferay.atlassian.net/browse/LPD-36796) - Modify history step creation logic for Color Picker, we only want it to happen on Blur
[LPD-34556](https://liferay.atlassian.net/browse/LPD-34556) - Sometimes Blur created an extra history step
[LPD-37459](https://liferay.atlassian.net/browse/LPD-37459) - Make sure history steps are synced between UndoRedo and data-engine

Could you review the commits under these ticket numbers? There is a pending PR at https://github.com/liferay-content-management/liferay-portal/pull/5901 However we need your approval as we made changes in data-engine.

Thank you,

### CODE REVIEW

**Comment by marcelabc** [comment]
Hey @rebsilva 
Could you review this PR, please? :smile:

**Comment by rebsilva** [comment]
@marcelabc sure! i will review this later today, thanks!

**Comment by rebsilva** [comment]
just started reviewing! :cherry_blossom:

**Comment by rebsilva** [comment]
i will already run the ci to see if any tests related to these changes fail

**Comment by rebsilva** [comment]
ci:test:data-engine

**Comment by rebsilva** [comment]
ci:test:forms

**Comment by rebsilva** [comment]
ci:test:playwright

**Comment by rebsilva** [comment]
ci:test:relevant

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Page.es.js [hunk 1]
```diff
@@ -177,6 +179,53 @@ const Page = ({
 		DDM_FORM_PORTLET_NAMESPACE
 	);
 
+	const {defaultLanguageId, history} = useFormState();
```
Comment (rebsilva) [inline modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Page.es.js:182]: why did you move the `useEffect` from the `WebContentVariant`? just want to know the reason behind it, since on `WebContentVariant` was more safe and clean since it only affects the web content page, while the Page file corresponds to all the features that use data engine

**Review by rebsilva — CHANGES_REQUESTED** [review]
hi @NemethNorbert, thanks a lot for this PR! i've made some questions and suggestions, please let me know what you think!

the tests are already running to see if they catch any unexpected behavior derived from these changes.

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Page.es.js (hunk 1 above)
Comment (NemethNorbert) [reply modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Page.es.js:182]: The web content variant was rendering the columns and we can have more than one columns in a given web content attaching more event listeners than needed. Moving these event listeners to the Page component ensured we won't create duplicated event listeners.

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/WebContentVariant.es.js [hunk 2]
```diff
@@ -49,6 +49,12 @@ export const Column = forwardRef(
 		const defaultLanguageIdRef = useRef(defaultLanguageId);
 		const dispatch = useForm();
 
+		useEffect(() => {
+			if (history?.steps.length === 0) {
+				dispatch({type: EVENT_TYPES.HISTORY.ADD});
+			}
+		}, [history, dispatch]);
+
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/WebContentVariant.es.js]: I left it separetly as they are two different effect. The first one is about creating the inital step at the first render of the Page Component, while the other one is about reseting the History state when the default locale changes.
I would not want to combine them because of this.

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Page.es.js (hunk 1 above)
Comment (rebsilva) [reply modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/Page.es.js:182]: i see, thanks for the explanation!

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/WebContentVariant.es.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rebsilva) [reply modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/core/components/PageRenderer/WebContentVariant.es.js]: i understand your point, let's leave as it is then

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/test/js/DatePicker/DatePicker.es.js [hunk 3]
```diff
@@ -50,6 +50,21 @@ describe('DatePicker', () => {
...
+				locale="en_US"
+				localizedValue={localizedValue}
+				predefinedValue="2021-08-20"
+				value="2020-06-02"
+			/>
+		);
+
+		expect(screen.getByRole('textbox')).toHaveValue('06/02/2020');
+	});
+
```
[the remark this answers is not in the document]
Comment (rebsilva) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/test/js/DatePicker/DatePicker.es.js:67]: i think is better to create a playwright test as well, thanks!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/DatePicker/DatePicker.es.js [hunk 4]
```diff
@@ -118,6 +118,19 @@ export default function DatePicker({
...
+	useEffect(() => {
+		if (date.rawDate !== '') {
+			const formattedDate = moment(rawDate, serverFormat, true)
+				.locale(locale ?? defaultLanguageId)
+				.format(momentFormat);
+
+			setDate((previousState) => ({...previousState, formattedDate}));
+		}
+	}, [date, defaultLanguageId, locale, momentFormat, rawDate, serverFormat]);
+
```
[the remark this answers is not in the document]
Comment (rebsilva) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/DatePicker/DatePicker.es.js:133]: oh i see now, well, i will test manually to see if this affects other expected behaviors but let's hope not :crossed_fingers:

**Comment by rebsilva** [comment]
i will analyze the tests failures and test manually asap, hoping to comeback with news unitl tomorrow morning, thanks!

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/test/js/DatePicker/DatePicker.es.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/test/js/DatePicker/DatePicker.es.js]: I removed the commit

**Comment by rebsilva** [comment]
taking a look at the tests failures

**Comment by rebsilva** [comment]
ci:test:relevant

**Comment by rebsilva** [comment]
@NemethNorbert the tests on our side seem to be fine, there aren't related failures. a lot of journal-web tests are failing but i believe you can see if they are related or not. from our side its okay to forward

**Review by rebsilva — APPROVED** [review]
DDM side LGTM

**Comment by NemethNorbert** [comment]
Thanks @rebsilva ,

We will forward it from the content-management repo. Thank you
